### DISCUSSION TITLE
Different expand behavior between `clojure-mode` and `clojure-ts-mode`

### DISCUSSION THREAD

**Discussion by dalugm** [minad/tempel#129] [Q&A] [answered]
I'm trying to add template for both [clojure-mode] and [clojure-ts-mode], let's use `do` as an example here (same problem occurs when I tried to use `n> r>` in other templates):

```
clojure-mode clojure-ts-mode

(do "(do" n> r> ")")
```

In `clojure-mode`, it works as expected:

<img width="580" alt="图片" src="https://github.com/minad/tempel/assets/25452934/8cdbd05e-fbf0-4794-bee6-8cbe3bf87514">

While in `clojure-ts-mode`, the inserted region is not indented correctly:

<img width="580" alt="图片" src="https://github.com/minad/tempel/assets/25452934/e1afdba7-ac3f-4337-817c-4f061cf916e8">

I don't know what's causing it, any help would be appreciated, thx in advance...

---

edit:

If I tried to expand the template inside a sexp, let's say:

```clojure
; `[]` is where my point is
(if do[])

; Expand inside `if` works fine in `clojure-mode` and `clojure-ts-mode`
(if (do
      []))
```

[clojure-mode]: https://github.com/clojure-emacs/clojure-mode
[clojure-ts-mode]: https://github.com/clojure-emacs/clojure-ts-mode

**Comment by minad** [comment]
I assume the problem is caused by differing indentation behavior of clojure-mode and clojure-ts-mode. Avoid > to prevent automatic indentation. Iirc there have been similar questions about the indentation behavior. I suggest to search the issue tracker and discussion forum.

**Reply by dalugm** [reply]
I had searched and they are not the same.

All I had found about indent problem in the issues and discussion are all about the wrong indent behavior, like `indented too deep` or `indented twice`

The problem I had here is in `*-ts-mode`, the `n> r>` is not indented at all, while `*-mode` works as expected. (Just like my screenshot shows)

**Reply by minad** [reply, marked as answer]
My suggestion would be to manually indent with spaces if `>` doesn't do the right thing. Alternatively you could try to override the indentation functions. To be clear, this is not a Tempel issue. Tempel just uses the indentation functionality from the underlying mode, and you have to adapt your templates accordingly.

**Reply by dalugm** [reply]
`tempel` indent by using `indent-according-to-mode`, am I right?

The problem is, if I trigger `indent-according-to-mode` manually, it can indent correctly whether you're in `*-mode` or `*-ts-mode`. But in tempel, it can't work as expected in `*-ts-mode` by expanding template's `n> r>`.

Using space manually actually worked but it's a little ugly...

**Reply by minad** [reply]
> But in tempel, it can't work as expected in *-ts-mode by expanding template's n> r>.

Tempel just executes the insertion and indentation commands step by step. If you execute the exact same sequence manually, you should get the same result.

**Reply by dalugm** [reply]
I have a simple snippet for `clojure-ts-mode` to test

```
(test "(test" n> ")")
```

Does this mean the following steps in `tempel`?

1. (insert "(test")
2. (insert "\n")
3. (indent-according-to-mode)
4. (insert ")")

After execute the four steps manully, there will be an indent before the final `)`; While in `tempel`, there is no indent.

I guess I have to learn more elisp to get out what tempel doing when expand...

**Reply by minad** [reply]
> After execute the four steps manully, there will be an indent before the final ); While in tempel, there is no indent.

Maybe `clojure-ts-mode` hasn't updated the AST yet when `indent-according-to-mode` is executed? Otherwise I could not explain a difference.

**Reply by dalugm** [reply]
> Maybe clojure-ts-mode hasn't updated the AST yet when indent-according-to-mode is executed?

I'm afraid all `*-ts-mode` have this problem... I'm using `dart-ts-mode` and `js-ts-mode` these days and met this problem on both of them: the place to input is not indented at all.

If you had `js-ts-mode`, you can try the following snippet:

```
js-mode js-ts-mode
(fn "function " p "(" p ") {" n> r> n "}" >)
```

In `js-mode`, it can be indented correctly while `js-ts-mode` can't.

**Reply by minad** [reply]
Okay, I suggest you take it to the treesitter devs. Maybe they have an idea. Ideally this should not make a difference, since otherwise this hurts the ability to script Emacs. Did you try executing keyboard macros, executing the steps? Do they work? Or if you write a function executing the insert, indent, insert, ... steps, does that work?

**Reply by minad** [reply]
It seems that indentation is generally different in treesitter modes. When I type `function () {` in `js-ts-mode` the next line is not indented. This is really weird. I am sure someone stumbled over this problem before and this surely must have been discussed on emacs-devel at some point? Note that this happens when I type manually, so the problem is not limited to Tempel.

**Reply by dalugm** [reply]
Thanks for your patience minad!

Would you mind reporting this to emacs-devel? As I'm not an native English speaker, it is hard for me to explain a problem clearly in English (like this we had commuicated for almost 1h...) And according to your description, I guess this is a huge problem, an Emacs expert like you can effectively communicate with the Emacs devels ;)

It's ok if you don't want to waste time on this since you had pointed out a way for me, thx again!

**Reply by minad** [reply]
> Would you mind reporting this to emacs-devel?

No, I will not report this. I am busy enough with my packages. Your English is fine and as far as I can tell the author of the Treesitter integration in Emacs is Chinese. ;)

**Reply by dalugm** [reply]
OK, maybe I'll choose to communicate with tree-sitter's author directly... 😄

**Reply by minad** [reply]
> OK, maybe I'll choose to communicate with tree-sitter's author directly... 😄

Yes, this is what I would suggest. Hopefully [Yuan](https://github.com/casouri) can answer directly, without taking the detour via emacs-devel, and point out a simple solution.

**Reply by casouri** [reply]
The reason is simple, when tempel inserts the `(test` and newline, the parse tree is kind of incomplete, because there's no closing paren. And the simple indentation rules in clojure-ts-mode doesn't account for it. It's the same in js-ts-mode, where `function () {` doesn't indent—it lacks a closing bracket. I guess clojure-ts-mode and js-ts-mode should add some heuristic indentation rules to account for this.

**Reply by minad** [reply]
@casouri Thank you for chiming in here. Your explanation makes sense. I expected that. But I wonder if users haven't questioned this before, when all the ts-modes have been introduced, since this is quite a deviation in indentation behavior from the existing modes and maybe also from the behavior in other editors? Has such a special rule been discussed before? Or maybe the problem is not so pronounced since people often use electric pair mode (I use `insert-pair` instead)? Another possibility for treesitter would be to automatically complete the AST. If there is a dangling brace, implicitly close it and treat the AST as complete. Would that be possible?

**Reply by casouri** [reply]
> Or maybe the problem is not so pronounced since people often use electric pair mode (I use insert-pair instead)? 

That's my impression. More importantly the developers (me, major mode authors, etc) use electric-pair-mode and doesn't get bothered by this. Plus tree-sitter modes are still relatively niche.

> Has such a special rule been discussed before?

Yes, multiple times. The problem is that these kind of rules are tricky to implement. If you want to do it right you might end up writing more and more code for edge cases (this depends, of course, here I'm only talking about my experience writing such rules for c-ts-mode; it might be easier to do for other languages.)

> Another possibility for treesitter would be to automatically complete the AST. If there is a dangling brace, implicitly close it and treat the AST as complete. Would that be possible?

It's possible, but it's generally outside of our control. IMO special heuristics are more realistic.

**Reply by minad** [reply]
@casouri Thanks. For Tempel the sitatuion is pretty bad because of the different indentation behavior. This means that the indentation fields `n>`, `r>` and `>` in templates are broken, since indentation is executed for incomplete ASTs.

**Reply by minad** [reply]
Maybe as a workaround I could replace the calls to `indent-according-to-mode` with a customizable variable, such that they could at least be overridden. I could also implement an heuristic rule in Tempel, which indents by a fixed width in ts modes. Alternatively it may make more sense to have to have this in treesitter directly. But then, it is understandable if such special rules are undesirable because they are error prone.

**Reply by casouri** [reply]
IMHO tempel is programming to an interface and shouldn't have to worry about the other side of the interface. Major modes should fix it. We should probably report to clojure-ts-mode's author and see what they can do.

**Reply by casouri** [reply]
For a workaround users can as you said type out the spaces for the indentation manually.
